src/communication/channelmanager.cpp

```diff
@@ -5,27 +5,293 @@
+            }
+
+            CloseChannels();
+
+            if (auto err = WaitTimeout(); !err.IsNone()) {
+                return err;
+            }
+        }
+    });
+}
+
+Error ChannelManager::WaitTimeout()
+{
+    aos::UniqueLock lock {mMutex};
+
+    if (auto err = mCondVar.Wait(lock, cReconnectPeriod, [this] { return mClose; }); !err.IsNone()) {
+        return err;
+    }
+
     return ErrorEnum::eNone;
 }
 
+Error ChannelManager::HandleRead()
+{
+    while (true) {
+        {
+            aos::UniqueLock lock {mMutex};
+            mCondVar.Wait(lock, [this] { return mChannels.Size() > 0 || mClose; });
+            if (mClose) {
+                return ErrorEnum::eNone;
+            }
+        }
+
+        AosProtocolHeader header;
+        if (auto ret = mTransport->Read(&header, sizeof(AosProtocolHeader));
+            ret < static_cast<int>(sizeof(AosProtocolHeader))) {
+            return {ErrorEnum::eRuntime, "failed to read header"};
+        }
+
+        if (header.mDataSize > cReadBufferSize) {
+            return {ErrorEnum::eRuntime, "data size is too big"};
+        }
+
+        uint32_t port    = header.mPort;
+        uint32_t msgSize = header.mDataSize;
+        memset(mTmpReadBuffer, 0, cReadBufferSize);
+
+        LOG_DBG() << "Read channel CM: port=" << port << " size=" << msgSize;
+
+        size_t totalRead = 0;
+        while (totalRead < msgSize) {
+            size_t remainingSize = msgSize - totalRead;
+            size_t readSize      = aos::Min(remainingSize, cReadBufferSize);
+
+            if (auto ret = mTransport->Read(mTmpReadBuffer + totalRead, readSize); ret < static_cast<int>(readSize)) {
+                return {ErrorEnum::eRuntime, "failed to read data into local buffer"};
+            }
+
+            totalRead += readSize;
+        }
+
+        UniqueLock lock {mMutex};
+
+        auto [channel, err] = mChannels.At(port);
+        if (!err.IsNone()) {
+            LOG_ERR() << "Failed to find channel: err=" << err.Message();
+
+            continue;
+        }
+
+        size_t processedSize = 0;
+        while (processedSize < msgSize) {
+            void*  buffer = nullptr;
+            size_t size   = 0;
+
+            if (auto err = channel->WaitRead(&buffer, &size); !err.IsNone()) {
+                LOG_ERR() << "Failed to wait read: err=" << err.Message();
+
+                break;
+            }
+
+            LOG_DBG() << "Request read channel CM: port=" << port << " size=" << size;
+
+            size = aos::Min(size, static_cast<size_t>(msgSize - processedSize));
+
+            memcpy(buffer, mTmpReadBuffer + processedSize, size);
+
+            if (auto err = channel->ReadDone(size); !err.IsNone()) {
+                LOG_ERR() << "Failed to read done: err=" << err.Message();
+
+                break;
+            }
+
+            processedSize += size;
+        }
+
+        LOG_DBG() << "Read channel done CM: port=" << port << " size=" << msgSize;
+    }
+}
```

Comment:
Could you divide this function into two? like read and process. or move process to separate function because HandleRead is to big.

Reply:
Done